File: core/core-api/src/test/java/com/coffee/gu/api/controller/RestDocsTest.java
package com.coffee.gu.api.controller;

import com.coffee.gu.auth.PrincipalArgumentResolver;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.restdocs.RestDocumentationContextProvider;
import org.springframework.restdocs.RestDocumentationExtension;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import static org.springframework.restdocs.mockmvc.MockMvcRestDocumentation.documentationConfiguration;

@ExtendWith(RestDocumentationExtension.class)
public abstract class RestDocsTest {

    protected MockMvc mockMvc;

    @BeforeEach
    void setUp(RestDocumentationContextProvider restDocumentation) {
        this.mockMvc = MockMvcBuilders.standaloneSetup(getController())
                .setCustomArgumentResolvers(new PrincipalArgumentResolver())
                .setControllerAdvice(new ApiControllerAdvice())
                .apply(documentationConfiguration(restDocumentation))
                .build();
    }

    protected abstract Object getController();
}

File: core/core-api/src/test/kotlin/com/coffee/gu/api/controller/RestDocsTest.kt
package com.coffee.gu.api.controller

import com.coffee.gu.auth.PrincipalArgumentResolver
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.restdocs.RestDocumentationContextProvider
import org.springframework.restdocs.RestDocumentationExtension
import org.springframework.restdocs.mockmvc.MockMvcRestDocumentation.documentationConfiguration
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.setup.MockMvcBuilders

@ExtendWith(RestDocumentationExtension::class)
abstract class RestDocsTest {

    lateinit var mockMvc: MockMvc

    @BeforeEach
    fun setUp(restDocumentation: RestDocumentationContextProvider) {
        val builder = MockMvcBuilders.standaloneSetup(controller)
            .setCustomArgumentResolvers(PrincipalArgumentResolver())
            .setControllerAdvice(ApiControllerAdvice())
        builder.apply<org.springframework.test.web.servlet.setup.StandaloneMockMvcBuilder>(documentationConfiguration(restDocumentation))
        this.mockMvc = builder.build()
    }

    protected abstract val controller: Any
}
